packages/compiler/core/checker.ts

```diff
@@ -3219,6 +3252,21 @@ export function createChecker(program: Program): Checker {
     return type;
   }
 
+  function resolveTypeReferenceString(reference: string): Type | undefined {
+    const parser = createParser(reference);
+    const node = parser.parseStandaloneReferenceExpression();
+    if (parser.parseDiagnostics.length > 0) {
+      program.reportDiagnostics(parser.parseDiagnostics);
+      return undefined;
+    }
+    const binder = createBinder(program);
+    binder.bindNode(node);
+    mutate(node).parent = globalNamespaceNode;
+
+    const type = checkTypeReference(node, undefined, false);
+    return type === errorType ? undefined : type;
+  }
```

Comment:
I'm curious how much this all costs. It's doing things we used to do only once per file and maybe could end up being called a lot in an emitter. We could always optimize later if it ends up in someone's hot path as a bottleneck. TS reuses a single parser for example.

Reply:
I think it's OK to wait and see on the perf, but on the other comment, we're now exposing a public API where you can call parse twice with reused state, so it would be a correctness issue if that doesn't work. Maybe we should instead just expose `parseStandaloneTypeReference(): [TypeReferenceNode, Diagnsotic[]]` rather than expose a usage that we don't use or test?

Reply:
updated to not expose the `createParser` function to keep in internal and expose that `parseStandaloneTypeReference` function instead